dotnet/Razorvine.Pyrolite/Pyrolite/Pickle/ArrayReader.cs

```diff
@@ -0,0 +1,56 @@
+﻿using System;
+
+namespace Razorvine.Pickle
+{
+    internal struct ArrayReader : IInputReader
```

Comment:
Why have an ArrayReader and a ReadOnlyMemoryReader? Can't we just always use the latter?

Reply:
The reason is that `ReadOnlyMemory` is fast only on .NET Core 3.0, while Array is fast for all runtimes. 

```cmd
dotnet run -c Release -- --filter *Memory* --runtimes netcoreapp2.0 netcoreapp2.1 netcoreapp3.0
```

|                     Method |     Toolchain | Count |       Mean |      Error |    StdDev | Ratio | RatioSD |   Gen 0 |  Gen 1 | Gen 2 | Allocated |
|--------------------------- |-------------- |------ |-----------:|-----------:|----------:|------:|--------:|--------:|-------:|------:|----------:|
|  DoublesFromReadOnlyMemory | netcoreapp2.0 |   100 |   8.258 us |  0.3034 us | 0.0166 us |  1.00 |    0.00 |  1.5411 | 0.0305 |     - |   9.57 KB |
|  DoublesFromReadOnlyMemory | netcoreapp2.1 |   100 |   6.170 us |  0.1856 us | 0.0102 us |  0.75 |    0.00 |  1.5488 | 0.0305 |     - |   9.57 KB |
|  DoublesFromReadOnlyMemory | netcoreapp3.0 |   100 |   4.779 us |  0.3199 us | 0.0175 us |  0.58 |    0.00 |  1.5488 | 0.0381 |     - |   9.52 KB |
|                            |               |       |            |            |           |       |         |         |        |       |           |
| IntegersFromReadOnlyMemory | netcoreapp2.0 |   100 |   7.069 us |  0.0940 us | 0.0052 us |  1.00 |    0.00 |  1.4877 | 0.0305 |     - |   9.18 KB |
| IntegersFromReadOnlyMemory | netcoreapp2.1 |   100 |   6.486 us |  0.3260 us | 0.0179 us |  0.92 |    0.00 |  1.4877 | 0.0305 |     - |   9.18 KB |
| IntegersFromReadOnlyMemory | netcoreapp3.0 |   100 |   4.579 us |  0.0578 us | 0.0032 us |  0.65 |    0.00 |  1.4877 | 0.0229 |     - |   9.13 KB |
|                            |               |       |            |            |           |       |         |         |        |       |           |
| BooleansFromReadOnlyMemory | netcoreapp2.0 |   100 |   3.300 us |  0.4791 us | 0.0263 us |  1.00 |    0.00 |  0.5417 | 0.0038 |     - |   3.35 KB |
| BooleansFromReadOnlyMemory | netcoreapp2.1 |   100 |   2.988 us |  0.2432 us | 0.0133 us |  0.91 |    0.01 |  0.5417 | 0.0038 |     - |   3.35 KB |
| BooleansFromReadOnlyMemory | netcoreapp3.0 |   100 |   2.023 us |  0.1900 us | 0.0104 us |  0.61 |    0.00 |  0.5455 | 0.0038 |     - |   3.34 KB |
|                            |               |       |            |            |           |       |         |         |        |       |           |
|  StringsFromReadOnlyMemory | netcoreapp2.0 |   100 |  17.708 us |  2.9890 us | 0.1638 us |  1.00 |    0.00 |  2.6550 | 0.1221 |     - |  16.53 KB |
|  StringsFromReadOnlyMemory | netcoreapp2.1 |   100 |  13.345 us |  0.2190 us | 0.0120 us |  0.75 |    0.01 |  2.6855 | 0.1221 |     - |  16.53 KB |
|  StringsFromReadOnlyMemory | netcoreapp3.0 |   100 |   9.514 us |  2.7487 us | 0.1507 us |  0.54 |    0.00 |  2.6245 | 0.1221 |     - |  16.13 KB |
|                            |               |       |            |            |           |       |         |         |        |       |           |
|  DoublesFromReadOnlyMemory | netcoreapp2.0 |  1000 |  72.713 us | 13.9506 us | 0.7647 us |  1.00 |    0.00 | 12.3291 | 1.9531 |     - |  76.37 KB |
|  DoublesFromReadOnlyMemory | netcoreapp2.1 |  1000 |  52.822 us |  5.1099 us | 0.2801 us |  0.73 |    0.00 | 12.3901 | 2.0142 |     - |  76.37 KB |
|  DoublesFromReadOnlyMemory | netcoreapp3.0 |  1000 |  39.876 us |  2.2336 us | 0.1224 us |  0.55 |    0.01 | 12.4512 | 2.0752 |     - |  76.32 KB |
|                            |               |       |            |            |           |       |         |         |        |       |           |
| IntegersFromReadOnlyMemory | netcoreapp2.0 |  1000 |  71.147 us | 20.6080 us | 1.1296 us |  1.00 |    0.00 | 11.7188 | 1.5869 |     - |  72.46 KB |
| IntegersFromReadOnlyMemory | netcoreapp2.1 |  1000 |  52.064 us |  0.8314 us | 0.0456 us |  0.73 |    0.01 | 11.7188 | 1.4648 |     - |  72.46 KB |
| IntegersFromReadOnlyMemory | netcoreapp3.0 |  1000 |  38.159 us |  4.2919 us | 0.2353 us |  0.54 |    0.01 | 11.7798 | 1.8921 |     - |  72.41 KB |
|                            |               |       |            |            |           |       |         |         |        |       |           |
| BooleansFromReadOnlyMemory | netcoreapp2.0 |  1000 |  28.035 us |  1.0109 us | 0.0554 us |  1.00 |    0.00 |  3.9673 | 0.2441 |     - |  24.45 KB |
| BooleansFromReadOnlyMemory | netcoreapp2.1 |  1000 |  25.773 us |  4.6881 us | 0.2570 us |  0.92 |    0.01 |  3.9673 | 0.2441 |     - |  24.45 KB |
| BooleansFromReadOnlyMemory | netcoreapp3.0 |  1000 |  16.309 us |  1.7685 us | 0.0969 us |  0.58 |    0.00 |  3.9673 | 0.2747 |     - |  24.45 KB |
|                            |               |       |            |            |           |       |         |         |        |       |           |
|  StringsFromReadOnlyMemory | netcoreapp2.0 |  1000 | 182.415 us | 30.4485 us | 1.6690 us |  1.00 |    0.00 | 25.6348 | 8.5449 |     - | 159.14 KB |
|  StringsFromReadOnlyMemory | netcoreapp2.1 |  1000 | 132.331 us | 36.0892 us | 1.9782 us |  0.73 |    0.02 | 25.6348 | 8.5449 |     - | 159.14 KB |
|  StringsFromReadOnlyMemory | netcoreapp3.0 |  1000 |  96.668 us | 32.8850 us | 1.8025 us |  0.53 |    0.01 | 25.2686 | 8.3008 |     - | 155.23 KB |